netquery/manager.go

```diff
@@ -0,0 +1,225 @@
+			}
+		}
+	}
+}
+
+func (mng *Manager) pushConnUpdate(ctx context.Context, conn Conn) error {
+	blob, err := json.Marshal(conn)
+	if err != nil {
+		return fmt.Errorf("failed to marshal connection: %w", err)
+	}
+
+	key := fmt.Sprintf("%s:%s%s", mng.runtimeReg.DatabaseName(), mng.pushPrefix, conn.ID)
+	wrapper, err := record.NewWrapper(
+		key,
+		new(record.Meta),
+		dsd.JSON,
+		blob,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create record wrapper: %w", err)
+	}
+
+	// FIXME(ppacher): it may happen that started != now for NEW connections.
+	// In that case we would push and UPD rather than NEW even if
+	// the connection is new ...
+	// Though, that's still better than always pushing NEW for existing
+	// connections.
+	// If we would use UnixNano() here chances would be even worse.
+	//
+	// Verify if the check in portbase/api/database.go is vulnerable
+	// to such timing issues in general.
+	wrapper.SetMeta(&record.Meta{
+		Created:  conn.Started.Unix(),
+		Modified: time.Now().Unix(),
+	})
+
+	mng.push(wrapper)
+	return nil
+}
+
+// convertConnection converts conn to the local representation used
+// to persist the information in SQLite. convertConnection attempts
+// to lock conn and may thus block for some time.
+func convertConnection(conn *network.Connection) (*Conn, error) {
+	conn.Lock()
+	defer conn.Unlock()
+
+	c := Conn{
+		ID:         genConnID(conn),
+		External:   conn.External,
+		IPVersion:  conn.IPVersion,
+		IPProtocol: conn.IPProtocol,
+		LocalIP:    conn.LocalIP.String(),
+		LocalPort:  conn.LocalPort,
+		Verdict:    conn.Verdict,
+		Started:    time.Unix(conn.Started, 0),
+		Tunneled:   conn.Tunneled,
+		Encrypted:  conn.Encrypted,
+		Internal:   conn.Internal,
+		Inbound:    conn.Inbound,
+		Type:       ConnectionTypeToString[conn.Type],
+	}
+
+	if conn.Ended > 0 {
+		ended := time.Unix(conn.Ended, 0)
+		c.Ended = &ended
+	}
+
+	extraData := map[string]interface{}{}
+
+	if conn.Entity != nil {
+		extraData["cname"] = conn.Entity.CNAME
+		extraData["blockedByLists"] = conn.Entity.BlockedByLists
+		extraData["blockedEntities"] = conn.Entity.BlockedEntities
+		extraData["reason"] = conn.Reason
+
+		c.RemoteIP = conn.Entity.IP.String()
+		c.RemotePort = conn.Entity.Port // FIXME(ppacher): or do we want DstPort() here?
+		c.Domain = conn.Entity.Domain
+		c.Country = conn.Entity.Country
+		c.ASN = conn.Entity.ASN
+		c.ASOwner = conn.Entity.ASOrg
+		c.Scope = conn.Entity.IPScope
+		if conn.Entity.Coordinates != nil {
+			c.Latitude = conn.Entity.Coordinates.Latitude
+			c.Longitude = conn.Entity.Coordinates.Longitude
+		}
+	}
+
+	// pre-compute the JSON blob for the extra data column
+	// and assign it.
+	extraDataBlob, err := json.Marshal(extraData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal extra data: %w", err)
+	}
+	c.ExtraData = extraDataBlob
+
+	return &c, nil
+}
+
+func genConnID(conn *network.Connection) string {
+	data := conn.ID + "-" + time.Unix(conn.Started, 0).String()
```

Comment:
Should we use the date first, so it would automatically sort them chronologically should this data land in a NoSQL db at some point?
Also, I would use a proper format for the date string.